Adds command to toggle cell outputs

JasonWeill opened the pull request:
<!--
Thanks for contributing to JupyterLab!
Please fill out the following items to submit a pull request.
See the contributing guidelines for more information:
https://github.com/jupyterlab/jupyterlab/blob/main/CONTRIBUTING.md
-->

## References

Preparatory work for https://github.com/jupyter/notebook/issues/7366

## Code changes

Adds a new command in the `notebook-extension`, `notebook:toggle-cell-outputs`, to toggle the cell outputs for the selected cells. If at least one output is shown, the new command hides all outputs. If no outputs of the selected cells are shown, it shows all outputs for the selected cells (if any).

## User-facing changes

I can bind this to a shortcut key `o` using the following JSON configuration in the Keyboard Shortcuts extension:

```javascript
{
    "shortcuts": [
        {
            "args": {},
            "command": "notebook:toggle-cell-outputs",
            "keys": [
                "O"
            ],
            "selector": ".jp-Notebook.jp-mod-commandMode"
        }
    ]
}
```

And then, as requested in https://github.com/jupyter/notebook/issues/7366, `o` toggles output visibility, for parity with nbclassic / Notebook 6.

## Backwards-incompatible changes

None.

krassowski commented on packages/notebook-extension/src/index.ts:
```diff
@@ -3357,6 +3359,29 @@ function addCommands(
     isEnabled
   });
+  commands.addCommand(CommandIDs.toggleOutput, {
+    label: trans.__('Toggle Visibility of Selected Outputs'),
```
Strictly: "of outputs of selected cell" (but this would be a bit too long probably)

krassowski commented on packages/notebook-extension/src/index.ts:
```diff
@@ -3357,6 +3359,29 @@ function addCommands(
+          if (notebook.isSelectedOrActive(cell) && cell.model.type === 'code') {
+            if ((cell as CodeCell).outputHidden === false) {
+              // We found at least one visible output; hide outputs for this cell
+              return NotebookActions.hideOutput(notebook);
```
Ah it relies on `hideOutput` behaviour of looping through all cells. I don't know how I feel about this. While code reuse is nice, I do not necessarily like that:
- a) if `NotebookActions.hideOutput` behaviour gets changed this function will stop working correctly - and there are no tests in this PR so we would not catch it
- b) we are iterating the notebook twice

but neither of these is a show stopper as (a) is unlikely to happen and (b) is unlikely to be a problem except for very very long notebooks

JasonWeill commented:
I added some unit test coverage for the show, hide, and toggle output functions, and refactored all three of these functions to be in the `NotebookActions` namespace.

krassowski reviewed (approved):
LGTM, thanks @JasonWeill!

krassowski commented:
@JasonWeill did you intend this to be included in 4.3 or 4.4? FYI AFAIK main branch is still at 4.3

krassowski commented:
We can branch out or revert but need to coordinate :)

JasonWeill commented:
I realized this just as I was adding a comment on the related notebook issue. Although it doesn't fix a release blocker, I think this change is low-risk enough to add to 4.3.0. OK to keep in 4.3.0?

krassowski commented:
No objections from my side :)

jtpio commented:
Maybe for the next times we could convert to draft the PRs set for the next milestone, so it's a bit more clear they should be merged later even if already approved.

JasonWeill commented:
@jtpio Or we could declare a 4.3.0 release branch when we're in the beta phase, so that `main` becomes the branch for future releases. Might be good to discuss on Wednesday.

jtpio commented:
Yes that could also work :+1:
